rust/src/protocol/mod.rs

```diff
@@ -1,166 +1,8 @@
-    }
-}
-
-#[derive(Error, Debug)]
-pub enum StateError {
-    #[error("state failed: channel error: {0}")]
-    ChannelError(&'static str),
-    #[error("state failed: round error: {0}")]
-    RoundError(#[from] RoundFailed),
-}
-
-pub struct PhaseState<S> {
-    // Inner state
-    inner: S,
-    // Coordinator state
-    coordinator_state: CoordinatorState,
-    // Request receiver halve
-    request_rx: mpsc::UnboundedReceiver<Request>,
-}
-
-// Functions that are available to all states
-impl<S> PhaseState<S> {
-    /// Receives the next [`Request`].
-    /// Returns [`StateError::ChannelError`] when all sender halve have been dropped.
-    async fn next_request(&mut self) -> Result<Request, StateError> {
-        debug!("received new message");
-        self.request_rx.recv().await.ok_or(StateError::ChannelError(
-            "all message senders have been dropped!",
-        ))
-    }
-
-    /// Handle an invalid request.
-    fn handle_invalid_message(response_tx: oneshot::Sender<Result<(), PetError>>) {
-        debug!("invalid message");
-        // `send` returns an error if the receiver halve has already been dropped. This means that
-        // the receiver is not interested in the response of the request. Therefore the error is
-        // ignored.
-        let _ = response_tx.send(Err(PetError::InvalidMessage));
-    }
-}
-
-#[derive(From)]
-pub enum StateMachine {
-    Idle(PhaseState<Idle>),
-    Sum(PhaseState<Sum>),
-    Update(PhaseState<Update>),
-    Sum2(PhaseState<Sum2>),
-    Unmask(PhaseState<Unmask>),
-    Error(PhaseState<StateError>),
-    Shutdown(PhaseState<Shutdown>),
-}
-
-impl StateMachine {
-    /// Move to the next state and consume the old one.
-    pub async fn next(self) -> Option<Self> {
-        match self {
-            StateMachine::Idle(state) => state.next().await,
-            StateMachine::Sum(state) => state.next().await,
-            StateMachine::Update(state) => state.next().await,
-            StateMachine::Sum2(state) => state.next().await,
-            StateMachine::Unmask(state) => state.next().await,
-            StateMachine::Error(state) => state.next().await,
-            StateMachine::Shutdown(state) => state.next().await,
-        }
-    }
-
-    /// Create a new state machine with the initial state `Idle`.
-    /// Fails if there is insufficient system entropy to generate secrets.
-    pub fn new() -> Result<(mpsc::UnboundedSender<Request>, Self), InitError> {
-        // crucial: init must be called before anything else in this module
-        sodiumoxide::init().or(Err(InitError))?;
-
-        let (request_tx, request_rx) = mpsc::unbounded_channel::<Request>();
-        let coordinator_state = CoordinatorState {
-            seed: RoundSeed::generate(),
-            ..Default::default()
-        };
-
-        Ok((
-            request_tx,
-            PhaseState::<Idle>::new(coordinator_state, request_rx).into(),
-        ))
-    }
-}
+pub mod state_machine;
 
 #[cfg(test)]
```

Comment:
I guess I should move the tests to `src/protocol/state_machine.rs` now, correct @Robert-Steiner ?

Reply:
Yes, that would be nice.

Reply:
done.